src/rgw/rgw_op.cc

```diff
@@ -6904,143 +7066,18 @@ void RGWDeleteMultiObj::execute(optional_yield y)
-	      continue;
-      }
-
-      if ((identity_policy_res == Effect::Pass && e == Effect::Pass && !acl_allowed)) {
-	      send_partial_response(*iter, false, "", -EACCES);
-	      continue;
-      }
-    }
-
-    uint64_t obj_size = 0;
-    std::string etag;
-
-    if (!rgw::sal::Object::empty(obj.get())) {
-      RGWObjState* astate = nullptr;
-      bool check_obj_lock = obj->have_instance() && bucket->get_info().obj_lock_enabled();
-      const auto ret = obj->get_obj_state(this, &astate, s->yield, true);
-
-      if (ret < 0) {
-        if (ret == -ENOENT) {
-          // object maybe delete_marker, skip check_obj_lock
-          check_obj_lock = false;
-        } else {
-          // Something went wrong.
-          send_partial_response(*iter, false, "", ret);
-          continue;
-        }
-      } else {
-        obj_size = astate->size;
-        etag = astate->attrset[RGW_ATTR_ETAG].to_str();
-      }
-
-      if (check_obj_lock) {
-        ceph_assert(astate);
-        int object_lock_response = verify_object_lock(this, astate->attrset, bypass_perm, bypass_governance_mode);
-        if (object_lock_response != 0) {
-          send_partial_response(*iter, false, "", object_lock_response);
-          continue;
-        }
-      }
-    }
-
-    // make reservation for notification if needed
-    const auto versioned_object = s->bucket->versioning_enabled();
-    const auto event_type = versioned_object && obj->get_instance().empty() ?
-      rgw::notify::ObjectRemovedDeleteMarkerCreated :
-      rgw::notify::ObjectRemovedDelete;
-    std::unique_ptr<rgw::sal::Notification> res
-      = store->get_notification(obj.get(), s->src_object.get(), s, event_type);
-    op_ret = res->publish_reserve(this);
-    if (op_ret < 0) {
-      send_partial_response(*iter, false, "", op_ret);
-      continue;
-    }
-
-    obj->set_atomic();
-
-    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op = obj->get_delete_op();
-    del_op->params.versioning_status = obj->get_bucket()->get_info().versioning_status();
-    del_op->params.obj_owner = s->owner;
-    del_op->params.bucket_owner = s->bucket_owner;
-    del_op->params.marker_version_id = version_id;
-
-    op_ret = del_op->delete_obj(this, y);
-    if (op_ret == -ENOENT) {
-      op_ret = 0;
-    }
-
-    send_partial_response(*iter, obj->get_delete_marker(), del_op->result.version_id, op_ret);
-
-    // send request to notification manager
-    int ret = res->publish_commit(this, obj_size, ceph::real_clock::now(), etag, version_id);
-    if (ret < 0) {
-      ldpp_dout(this, 1) << "ERROR: publishing notification failed, with error: " << ret << dendl;
-      // too late to rollback operation, hence op_ret is not set here
+    rgw_obj_key* obj_key = &*iter;
+    if (y) {
+      spawn::spawn(y.get_yield_context(), [this, &y, obj_key] (yield_context yield) {
+        handle_individual_object(obj_key, optional_yield { y.get_io_context(), yield }); 
```

Comment:
my only concern here is unbounded concurrency - each request could spawn 1000 or more osd requests at a time

we're generally careful to limit concurrency so we don't disrupt the service of other requests. for example, we recently added concurrency to CopyObject in https://github.com/ceph/ceph/pull/48155 with a default limit of 10. ListObjects concurrency was raised from 8 to 128 in https://github.com/ceph/ceph/pull/28558

we've used the primitives in rgw_aio.h and rgw_aio_throttle.h to do this elsewhere, like in that recent CopyObject pr. you'd have to implement `handle_individual_object()` as an `Aio::OpFunc` similar to the `librados_op()` in rgw_aio.cc

another option is to build your own throttling logic around this `formatter_flush_cond`

Reply:
you can have a counter inside the loop, and call `wait_flush()` whenever you hit the limit.
it would not be exactly "I/O rate" but would ensure we don't have unbounded concurrency.
if the limit is set high enough (128?) it would not have a negative impact on the op itself.

Reply:
Thanks for the suggestion @yuvalif . I used a slightly modified version of that, where the stop criteria is passed in as a predicate so that I can keep up to `rgw_multi_obj_del_max_aio` coroutines active at all times.